File: tools/mph.py
#!/usr/bin/python
# Easy Perfect Minimal Hashing
#
# Based on:
# By Steve Hanov. Released to the public domain.
# http://stevehanov.ca/blog/index.php?id=119
#
# Based on:
# Edward A. Fox, Lenwood S. Heath, Qi Fan Chen and Amjad M. Daoud,
# "Practical minimal perfect hash functions for large databases",
# CACM, 35(1):105-121
#
# also a good reference:
# Compress, Hash, and Displace algorithm by Djamal Belazzougui,
# Fabiano C. Botelho, and Martin Dietzfelbinger

import sys
import time

PRIME = 0x01000193
INTERNAL_ENCODING = "UTF-8"


# Calculates a distinct hash function for a given string. Each value of the
# integer d results in a different hash value.
def hash(d, str):
	if d == 0:
		d = PRIME

	c = int(str, base=16)

	# It doesn't matter for MPH if it's FVN or not until G
	# is correctly filled, therefore simple XOR is enough to produce
	# required randomness while produced index fits into uint16_t.
	#
	# You can consider this as usage of Unicode codepoint as a hash
	# itself, but it need to depend on d to distibute values between
	# buckets

	return d ^ c


# Computes a minimal perfect hash table using the given python dictionary. It
# returns a tuple (G, V). G and V are both arrays. G contains the intermediate
# table of values needed to compute the index of the value in V. V contains the
# values of the dictionary.
def CreateMinimalPerfectHash(dict):
	size = len(dict)

	# Step 1: Place all of the keys into buckets
	buckets = [[] for i in range(size)]
	G = [0] * size
	values = [None] * size

	for key in dict.keys():
		buckets[hash(0, key) % size].append(key)

	# Step 2: Sort the buckets and process the ones with the most items first.
	buckets.sort(key=len, reverse=True)
	for b in xrange(size):
		bucket = buckets[b]
		if len(bucket) <= 1:
			break

		d = 1
		item = 0
		slots = []

		# Repeatedly try different values of d until we find a hash function
		# that places all items in the bucket into free slots
		while item < len(bucket):
			slot = hash(d, bucket[item]) % size
			if values[slot] is not None or slot in slots:
				d += 1
				item = 0
				slots = []
			else:
				slots.append(slot)
				item += 1

		G[hash(0, bucket[0]) % size] = d
		for i in range(len(bucket)):
			values[slots[i]] = dict[bucket[i]]

	# Only buckets with 1 item remain. Process them more quickly by directly
	# placing them into a free slot. Use a negative value of d to indicate
	# this.
	freelist = []
	for i in xrange(size):
		if values[i] is None:
			freelist.append(i)

	for b in xrange(b, size):
		bucket = buckets[b]

		if len(bucket) == 0:
			break

		slot = freelist.pop()
		# We subtract one to ensure it's negative even if the zeroeth slot was
		# used.
		G[hash(0, bucket[0]) % size] = -slot - 1
		values[slot] = dict[bucket[0]]

	return (G, values)


# Look up a value in the hash table, defined by G and V.
def PerfectHashLookup(G, key):
	d = G[hash(0, key) % len(G)]
	if d < 0:
		return -d - 1
	return hash(d, key) % len(G)


# Print human-readable info regarding this hash-table
def GenerateInfo(G, combined):
	print '''/*
%d
PRIME: %08X,
G_SIZE: %d,
COMBINED_LENGTH: %d,
ENCODING: %s,
*/
''' % (time.time(),
		PRIME, len(G),
		len(combined) / 4,
		INTERNAL_ENCODING)


VALUE_TEMPLATE = '''	{ 0x%(codepoint)05X, %(decomps)s },'''
VALUE_REF_TEMPLATE = '''&V%(codepoint)05X'''


# Filter out non-characters from private area
def Pass(c):
	assert(len(c) > 0)
	return (ord(c) < 0xE000 or ord(c) > 0xF8FF)


# Produce C-source-ready decomposition string
def FormatReplacement(r):
	chars = u''.join(
		filter(Pass, [unichr(int(x, base=16)) for x in r])).encode(INTERNAL_ENCODING)
	if len(chars) == 0:
		return
	formatted = '%s' % (''.join(('\\x%02X' % (ord(x)) for x in chars)))
	return formatted


# Print values table
def GenerateValues(G, V):
	BOUNDARY = 8

	print '/* codepoints */'
	print 'static const uint32_t VALUES_C[] = {'
	for i, (codepoint, replacement) in enumerate(V):
		assert(replacement is not None)

		if i % BOUNDARY == 0:
			sys.stdout.write('\t')
		sys.stdout.write('0x%06X, ' % (int(codepoint, base=16),))
		if (i + 1) % BOUNDARY == 0:
			sys.stdout.write('\n')

	print '};'
	print

	BOUNDARY = 10

	print '/* indexes */'
	print 'static const uint16_t VALUES_I[] = {'
	for i, (codepoint, replacement) in enumerate(V):
		assert(replacement is not None)

		if i % BOUNDARY == 0:
			sys.stdout.write('\t')
		sys.stdout.write('0x%04X, ' % (replacement,))
		if (i + 1) % BOUNDARY == 0:
			sys.stdout.write('\n')

	print '};'
	print


# Print first hash table
def GenerateG(G):
	BOUNDARY = 12

	print 'static const int16_t G[] = {'
	for i, x in enumerate(G):
		if i % BOUNDARY == 0:
			sys.stdout.write('\t')
		sys.stdout.write('%d' % (x,))
		sys.stdout.write(', ')
		if (i + 1) % BOUNDARY == 0:
			sys.stdout.write('\n')
	print '};'
	print


# Print combined encoded string
def GenerateCombined(combined):
	BOUNDARY = 12 * 4

	def chunks(combined, n):
		for i in xrange(0, len(combined), n):
			yield combined[i:i + n]

	print 'static const uint8_t COMBINED[] = {'
	for chunk in chunks(combined, BOUNDARY):
		parts = chunks(chunk, 4)
		# replace \xYY with 0xYY
		print '\t%s,' % (', '.join((p.replace('\\', '0') for p in parts)))
	print '};'
	print

File: tools/mph.go
package main

import (
	"sort"
)

// Constants common for entire MPH generation
const (
	MPHPrime            = 0x01000193 // No particular reason, because FVN is not currently used. Should match prime used in _nu_hash().
	MPHInternalEncoding = "UTF-8"    // Internal encoding for COMBINED
)

// MPHGType : G array type
type MPHGType []int16

// MPHV : internal type used to build MPH
type MPHV struct {
	codepoint   uint32 // Perhaps should have been `rune`, but for now keep this as close to generated C code as possible
	replacement uint16
}

// MPHVType : internal type used during MPH creation,
// C and I are coming out of this array, codepoint goes into C,
// replacement goes into I
type MPHVType []MPHV

// MPHCType : codepoints (C) array type
type MPHCType []uint32

// MPHIType : indices (I) array type
type MPHIType []uint16

// MPHCombinedType : COMBINED array type
// This array is build by mph-combined
type MPHCombinedType []uint8

// MPHMapping : source of MPH build
type MPHMapping map[uint32]MPHV

// Calculates a distinct hash function for a given string.
// Each value of the integer d results in a different hash value.
//
// Should loosely match _nu_hash()
func hash(d uint32, codepoint uint32) uint32 {
	if d == 0 {
		d = MPHPrime
	}

	// It doesn't matter for MPH if it's FVN or not until G
	// is correctly filled, therefore simple XOR is enough to produce
	// required randomness while produced index fits into uint16_t.
	//
	// You can consider this as usage of Unicode codepoint as a hash
	// itself, but it need to depend on d to distibute values between
	// buckets

	return (d ^ codepoint)
}

// Easy Perfect Minimal Hashing
//
// Based on:
// By Steve Hanov. Released to the public domain.
// http://stevehanov.ca/blog/index.php?id=119
//
// Based on:
// Edward A. Fox, Lenwood S. Heath, Qi Fan Chen and Amjad M. Daoud,
// "Practical minimal perfect hash functions for large databases",
// CACM, 35(1):105-121
//
// Also a good reference:
// Compress, Hash, and Displace algorithm by Djamal Belazzougui,
// Fabiano C. Botelho, and Martin Dietzfelbinger
//
// ---
//
// Computes a minimal perfect hash table using the given python dictionary.
// It returns a tuple (G, V). G and V are both arrays. G contains the
// intermediate values needed to compute the index of the value in V.
// V contains the values of the incoming dictionary.
func createMPH(mapping MPHMapping) (MPHGType, MPHVType) {
	size := uint32(len(mapping))

	// Step 1: Place all keys into buckets
	buckets := make([][]uint32, size)
	G := make(MPHGType, size)
	V := make(MPHVType, size)

	for key := range mapping {
		bucket := (hash(0, key) % size)
		buckets[bucket] = append(buckets[bucket], key)
	}

	// Step 2: Sort the buckets and process the ones with the most items first
	sort.Slice(buckets, func(i, j int) bool {
		return (len(buckets[i]) > len(buckets[j]))
	})

	lastBucket := 0
	for i, bucket := range buckets {
		// Stop this stage when reaching bucket with length of 1 or 0.
		// Since buckets are sorted, tail of buckets has either buckets with
		// length 1 or 0 (as step 3 expecpects).
		if len(bucket) <= 1 {
			lastBucket = i
			break
		}

		var d uint32 = 1
		slots := []uint32{}

		// Repeatedly try different values of d until we find a hash function
		// that places all items in the bucket into free slots
		for item := 0; item < len(bucket); {
			slot, slotBusy := hash(d, bucket[item])%size, false

			for _, existingSlot := range slots {
				if existingSlot == slot {
					slotBusy = true
					break
				}
			}

			// If slot is already taken, or codepoint is assigned in V,
			// reset item and clear slots, then try another d
			if V[slot].codepoint != 0 || slotBusy {
				item, slots = 0, slots[:0]
				d++
			} else {
				slots = append(slots, slot)
				item++
			}
		}

		G[hash(0, bucket[0])%size] = int16(d)
		for i := 0; i < len(bucket); i++ {
			V[slots[i]] = mapping[bucket[i]]
		}
	}

	// Step 3: Only buckets with 1 item remain. process them more quickly by directly
	// placing them into a free slot. Use a negative value of d to indicate this.
	freelist := []uint32{}
	for i := uint32(0); i < size; i++ {
		if V[i].codepoint == 0 {
			freelist = append(freelist, i)
		}
	}

	for i := lastBucket; i < len(buckets); i++ {
		bucket := buckets[i]
		if len(bucket) < 1 {
			break
		}

		slot := freelist[i-lastBucket]

		// We subtract one to ensure it's negative even if the zeroeth slot was used
		G[hash(0, bucket[0])%size] = int16(-slot - 1)
		V[slot] = mapping[bucket[0]]
	}

	return G, V
}

// Splits generated V into C and I arrays
func splitCodepointsAndIndices(v MPHVType) (MPHCType, MPHIType) {
	size := len(v)
	C := make(MPHCType, size)
	I := make(MPHIType, size)

	for i, entry := range v {
		C[i], I[i] = entry.codepoint, entry.replacement
	}

	return C, I
}
